### PR TITLE
raftstore: add commit log duration metrics

### PR DESCRIPTION
**Pull request by Connor1996**
## What have you changed? (mandatory)
add commit log duration metrics which spans from propose to committed

## What are the type of the changes? (mandatory)
- Improvement (non-breaking change which is an improvement to an existing feature)

## How has this PR been tested? (mandatory)
manual test

## Does this PR affect documentation (docs/docs-cn) update? (mandatory)
No

## Does this PR affect tidb-ansible update? (mandatory)
https://github.com/pingcap/tidb-ansible/pull/577

### CODE REVIEW

File: src/raftstore/store/peer.rs [hunk 1]
```diff
@@ -1037,12 +1042,16 @@ impl Peer {
...
+            let t = Instant::now();
             for entry in committed_entries.iter().rev() {
                 // raft meta is very small, can be ignored.
                 self.raft_log_size_hint += entry.get_data().len() as u64;
                 if lease_to_be_updated {
                     let propose_time = self.find_propose_time(entry.get_index(), entry.get_term());
                     if let Some(propose_time) = propose_time {
+                        raft_metrics
+                            .commit_log
+                            .observe(duration_to_sec(t.elapsed()));
```
Comment (siddontang) [inline src/raftstore/store/peer.rs]: em, not t - propose_time?

Comment (Connor1996) [reply src/raftstore/store/peer.rs]: oops, mistakes when rewriting it

Comment (siddontang) [reply src/raftstore/store/peer.rs]: oh, the result is right in your test 😭 ?

Comment (Connor1996) [reply src/raftstore/store/peer.rs]: yes, I make the rewrite to use local metrics after testing

File: src/raftstore/store/peer.rs [hunk 2]
```diff
@@ -1037,12 +1042,16 @@ impl Peer {
                 // have no effect.
                 self.proposals.clear();
             }
+            let now = monotonic_raw_now();
```
Comment (overvenus) [inline src/raftstore/store/peer.rs]: It is expensive. Can we sample instead?

Comment (Connor1996) [reply src/raftstore/store/peer.rs]: maybe we can just observe one committed_entry for every ready_apply, is that acceptable?

Comment (Connor1996) [reply src/raftstore/store/peer.rs]: maybe we can just observe one committed_entry for every ready_apply, is that acceptable?

File: src/raftstore/store/peer.rs [hunk 3]
```diff
@@ -1037,12 +1042,16 @@ impl Peer {
...
             for entry in committed_entries.iter().rev() {
                 // raft meta is very small, can be ignored.
                 self.raft_log_size_hint += entry.get_data().len() as u64;
                 if lease_to_be_updated {
                     let propose_time = self.find_propose_time(entry.get_index(), entry.get_term());
                     if let Some(propose_time) = propose_time {
+                        raft_metrics
+                            .commit_log
+                            .observe(duration_to_sec((now - propose_time).to_std().unwrap()));
```
Comment (solotzg) [inline src/raftstore/store/peer.rs]: I just wonder why you didn't update `now` after using it.

Comment (Connor1996) [reply src/raftstore/store/peer.rs]: seems that it is unnecessary for time to be so precise.

File: src/raftstore/store/peer.rs (hunk 2 above)
Comment (overvenus) [reply src/raftstore/store/peer.rs]: TiKV invokes it frequently, so making a syscall is unacceptable. Instead, you can choose a less precise clock, like CLOCK_MONOTONIC_COARSE.

**Comment by BusyJay** [comment]
Please update your pull request.

**Pull request closed without merging**